### PR TITLE
PP-9226: Pipeline to build and push reverse-proxy image to ECR

### PR DESCRIPTION
**Pull request by katstevens**
The goal here is to replace the existing Jenkins job that does the following:
- searches the pay-infra repo for NAXSI config files
- builds the `reverse-proxy` image from the files in pay-scripts 
- copies the NAXSI files into the nginx config on the image (see [this helper script](https://github.com/alphagov/pay-scripts/blob/master/images/proxy/build-latest-master.sh))
- pushes the image to Dockerhub, for use in end-to-end tests

To replicate this, I've built a new pipeline (that will shortly include builds for some other E2E helper images), `e2e-helpers`. Here's what's included.

Resources:
- `reverse-proxy-src`: i.e. the pay-scripts repo. Changes to the `images/proxy` folder will trigger the pipeline job.
- `pay-infra-src`: i.e. the pay-infra repo. Solely used as a source of NAXSI config files.
- `reverse-proxy-ecr-registry-test`: the ECR repository we want to push to.
- `reverse-proxy-dockerhub`: the Dockerhub repository we also push to (for backwards compatibility)

Tasks:
- `find-and-compile-naxsi-rules`: takes both pay-infra and pay-scripts repos, finds the files in pay-infra and copies them into a new folder 'target' in pay-scripts, which is specified in the [Dockerfile](https://github.com/alphagov/pay-scripts/blob/master/images/proxy/Dockerfile#L19). Some wrangling of inputs and outputs means that the 'target' folder can be persisted along to the next task (hurray for the [Concourse docs](https://blog.concourse-ci.org/introduction-to-task-inputs-and-outputs/) that had an example of exactly what I wanted to do!)
- `build-reverse-proxy-image`: pretty much copied and pasted from our other image builds. The `CONTEXT` param points to the folder where the Dockerfile lives.
- `reverse-proxy-ecr-registry-test`: basic push to ECR, all the config for this is defined on the resource itself
- `reverse-proxy-dockerhub`: basic push to Dockerhub, all the config for this is defined on the resource itself

Tagging the image as `latest-master` for now - we've agreed to address versioned tagging in a later story.

The pipeline is visible on Concourse: https://pay-cd.deploy.payments.service.gov.uk/teams/pay-dev/pipelines/e2e-helpers. The `reverse-proxy` image built by this job has been tested in this e2e run: https://pay-cd.deploy.payments.service.gov.uk/teams/pay-dev/pipelines/pr-ci/jobs/endtoend-products-e2e/builds/97.

### CODE REVIEW

File: ci/pipelines/e2e-helpers.yml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    type: git
+    icon: github
+    source:
+      uri: https://github.com/alphagov/pay-scripts
+      branch: master
+      username: alphagov-pay-ci
+      password: ((github-access-token))
+      paths:
+        # Should trigger on any changes to this folder
+        - images/proxy/*
```
Comment (jfharden) [inline ci/pipelines/e2e-helpers.yml:13]: These are glob patters, are you happy for it only to be changes in images/proxy or do you need subfolders too (in which case you want `**` I believe. You might need **/* 🤔

File: ci/pipelines/e2e-helpers.yml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            - name: reverse-proxy-src
+          outputs:
+            - name: reverse-proxy-src
+          run:
+            path: /bin/sh
+            args:
+              - -ec
+              - |
+                cd reverse-proxy-src/images/proxy
+                ls -lrt
```
Comment (jfharden) [inline ci/pipelines/e2e-helpers.yml:76]: Do you want to leave this in (and the one at the end) or was it just while debugging?

File: ci/pipelines/e2e-helpers.yml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            type: registry-image
+            source:
+              repository: vito/oci-build-task
+          inputs:
+          - name: reverse-proxy-src
+          outputs:
+          - name: image
+          run:
+            path: build
+      - put: reverse-proxy-ecr-registry-test
```
Comment (jfharden) [inline ci/pipelines/e2e-helpers.yml:100]: I'd suggest for your last to puts to run them in paralell

```
      - in_parallel:
        - put: reverse-proxy-ecr-registry-test
          params:
            image: image/image.tar
        - put: reverse-proxy-dockerhub
          params:
            image: image/image.tar
```

**Review by jfharden — COMMENTED** [review]
Couple of questions but looks great overall :)

File: ci/pipelines/e2e-helpers.yml (hunk 1 above)
Comment (katstevens) [reply ci/pipelines/e2e-helpers.yml:13]: Just that (flat) folder, think we're ok https://github.com/alphagov/pay-scripts/tree/master/images/proxy

File: ci/pipelines/e2e-helpers.yml (hunk 2 above)
Comment (katstevens) [reply ci/pipelines/e2e-helpers.yml:76]: Good spot! Having both there was super useful when checking to see if the `target` folder had appeared or not. Will remove now though

File: ci/pipelines/e2e-helpers.yml (hunk 3 above)
Comment (katstevens) [reply ci/pipelines/e2e-helpers.yml:100]: Excellent suggestion! Will do

File: ci/pipelines/e2e-helpers.yml (hunk 1 above)
Comment (jfharden) [reply ci/pipelines/e2e-helpers.yml:13]: Cool 👍

**Pull request merged**